cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java

```diff
@@ -149,7 +159,62 @@ private Plugin findPlugin(String pluginType, String pluginName, String pluginId,
-                                       pluginType, pluginName, properties, selector);
+
+
+    final Class[] callerClasses = CallerClassSecurityManager.getCallerClasses();
+    if (callerClasses.length < 3) {
+      // This shouldn't happen as there should be someone calling this method.
+      throw new IllegalStateException("Invalid call stack.");
+    }
+
+    Set<ArtifactId> seenParents = new HashSet<>();
+    List<ArtifactId> parents = new ArrayList<>();
+    // go through the call stack to get the parent plugins. This is to support plugins of plugins.
+    // for example, suppose appX uses pluginY, which uses pluginZ. When we try to use pluginZ, we need to first
+    // check if pluginZ extends pluginY. If it does not, we then check if it extends appX.
+    // 0 is the CallerClassSecurityManager, 1 is this class, hence 2 is the actual caller
+    for (int i = 2; i < callerClasses.length; i++) {
+      ClassLoader callerClassloader = callerClasses[i].getClassLoader();
+      if (AbstractPluginConfigurable.class.isAssignableFrom(callerClasses[i]) || callerClassloader == null) {
+        // we've reached the program, no need to go higher up the call stack.
+        break;
+      } else if (callerClassloader instanceof PluginClassLoader) {
+        // if this is the first time we've seen this plugin artifact, it must be a new plugin parent.
+        ArtifactId pluginCallerArtifactId = ((PluginClassLoader) callerClassloader).getArtifactId();
+        if (seenParents.add(pluginCallerArtifactId)) {
+          parents.add(pluginCallerArtifactId);
+        }
+      }
+    }
+
+    ArtifactRange parentRange;
+    for (ArtifactId pluginParent : parents) {
+      Id.Artifact parentId = Id.Artifact.from(deployNamespace, pluginParent);
+      parentRange = new ArtifactRange(parentId.getNamespace().getId(), parentId.getName(),
+                                      parentId.getVersion(), true, parentId.getVersion(), true);
+      try {
+        Map.Entry<ArtifactDescriptor, PluginClass> pluginEntry =
+          artifactRepository.findPlugin(deployNamespace.toEntityId(), parentRange, pluginType, pluginName, selector);
+        return FindPluginHelper.getPlugin(parents, pluginEntry, properties, pluginType, pluginName, pluginInstantiator);
+      } catch (PluginNotExistsException e) {
+        // ignore this in case the plugin extends something higher up in the call stack.
+        // For example, suppose the app uses pluginA, which uses pluginB. However, pluginB is a plugin that
+        // has the app as its parent and not pluginA as its parent. In this case, we want to keep going up the call
+        // stack until we get to the app as the parent, which would be able to find plugin B.
+      } catch (IOException e) {
+        throw Throwables.propagate(e);
+      }
+    }
+
+    parentRange = new ArtifactRange(artifactId.getNamespace().getId(), artifactId.getName(),
```

Comment:
Seems like the code would be cleaner if you just add this `ArtifactRange` into `parents` after the first for-loop, such that we don't need this extra code block, but rather just throw the last encountered `PluginNotExistsException` if it reaches here.